File: playground/cpp/days/day02/tests/bounded_stack_test.cpp
#include "day02/bounded_stack.h"
#include <gtest/gtest.h>

TEST(BoundedStack, StartsEmpty) {
    BoundedStack s(2);
    EXPECT_TRUE(s.empty());
    EXPECT_EQ(s.size(), 0u);
    EXPECT_FALSE(s.full());
}

TEST(BoundedStack, PushUpToCapacity) {
    BoundedStack s(2);
    s.push(5);
    s.push(7);
    EXPECT_EQ(s.size(), 2u);
    EXPECT_TRUE(s.full());
}

TEST(BoundedStack, PushBeyondCapacityThrows) {
    BoundedStack s(1);
    s.push(1);
    EXPECT_THROW(s.push(2), std::runtime_error);
    EXPECT_EQ(s.size(), 1u);  // rejected push must not touch state
}

TEST(BoundedStack, PopReturnsLifoOrder) {
    BoundedStack s(3);
    s.push(1);
    s.push(2);
    s.push(3);
    EXPECT_EQ(s.pop(), 3);
    EXPECT_EQ(s.pop(), 2);
    EXPECT_EQ(s.size(), 1u);
}

TEST(BoundedStack, PopEmptyThrows) {
    BoundedStack s(1);
    EXPECT_THROW(s.pop(), std::runtime_error);
}

TEST(BoundedStack, TopDoesNotRemove) {
    BoundedStack s(2);
    s.push(9);
    EXPECT_EQ(s.top(), 9);
    EXPECT_EQ(s.size(), 1u);
}

TEST(BoundedStack, TopEmptyThrows) {
    BoundedStack s(1);
    EXPECT_THROW(s.top(), std::runtime_error);
}

TEST(BoundedStack, ZeroCapacityIsAlwaysFull) {
    BoundedStack s(0);
    EXPECT_TRUE(s.full());
    EXPECT_THROW(s.push(1), std::runtime_error);
}

TEST(BoundedStack, PopFreesCapacityForAnotherPush) {
    BoundedStack s(1);
    s.push(1);
    ASSERT_TRUE(s.full());
    s.pop();
    EXPECT_FALSE(s.full());
    EXPECT_TRUE(s.empty());
    s.push(2);  // must succeed -- capacity was freed, not stuck full forever
    EXPECT_EQ(s.top(), 2);
    EXPECT_EQ(s.size(), 1u);
}

TEST(BoundedStack, TopReflectsMostRecentPush) {
    BoundedStack s(3);
    s.push(1);
    EXPECT_EQ(s.top(), 1);
    s.push(2);
    EXPECT_EQ(s.top(), 2);
    s.push(3);
    EXPECT_EQ(s.top(), 3);
}

TEST(BoundedStack, EmptyAfterPoppingEverything) {
    BoundedStack s(2);
    s.push(1);
    s.push(2);
    s.pop();
    s.pop();
    EXPECT_TRUE(s.empty());
    EXPECT_EQ(s.size(), 0u);
    EXPECT_THROW(s.pop(), std::runtime_error);
}

File: playground/java/days/day02/src/test/java/day02/BoundedStackTest.java
package day02;

import org.junit.jupiter.api.Test;

import java.util.NoSuchElementException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class BoundedStackTest {

    @Test
    void startsEmpty() {
        BoundedStack s = new BoundedStack(2);
        assertTrue(s.isEmpty());
        assertEquals(0, s.size());
        assertFalse(s.isFull());
    }

    @Test
    void pushUpToCapacity() {
        BoundedStack s = new BoundedStack(2);
        s.push(5);
        s.push(7);
        assertEquals(2, s.size());
        assertTrue(s.isFull());
    }

    @Test
    void pushBeyondCapacityThrows() {
        BoundedStack s = new BoundedStack(1);
        s.push(1);
        assertThrows(IllegalStateException.class, () -> s.push(2));
        assertEquals(1, s.size()); // rejected push must not touch state
    }

    @Test
    void popReturnsLifoOrder() {
        BoundedStack s = new BoundedStack(3);
        s.push(1);
        s.push(2);
        s.push(3);
        assertEquals(3, s.pop());
        assertEquals(2, s.pop());
        assertEquals(1, s.size());
    }

    @Test
    void popEmptyThrows() {
        BoundedStack s = new BoundedStack(1);
        assertThrows(NoSuchElementException.class, s::pop);
    }

    @Test
    void topDoesNotRemove() {
        BoundedStack s = new BoundedStack(2);
        s.push(9);
        assertEquals(9, s.top());
        assertEquals(1, s.size());
    }

    @Test
    void topEmptyThrows() {
        BoundedStack s = new BoundedStack(1);
        assertThrows(NoSuchElementException.class, s::top);
    }

    @Test
    void zeroCapacityIsAlwaysFull() {
        BoundedStack s = new BoundedStack(0);
        assertTrue(s.isFull());
        assertThrows(IllegalStateException.class, () -> s.push(1));
    }

    @Test
    void popFreesCapacityForAnotherPush() {
        BoundedStack s = new BoundedStack(1);
        s.push(1);
        assertTrue(s.isFull());
        s.pop();
        assertFalse(s.isFull());
        assertTrue(s.isEmpty());
        s.push(2); // must succeed -- capacity was freed, not stuck full
        assertEquals(2, s.top());
        assertEquals(1, s.size());
    }

    @Test
    void topReflectsMostRecentPush() {
        BoundedStack s = new BoundedStack(3);
        s.push(1);
        assertEquals(1, s.top());
        s.push(2);
        assertEquals(2, s.top());
        s.push(3);
        assertEquals(3, s.top());
    }

    @Test
    void emptyAfterPoppingEverything() {
        BoundedStack s = new BoundedStack(2);
        s.push(1);
        s.push(2);
        s.pop();
        s.pop();
        assertTrue(s.isEmpty());
        assertEquals(0, s.size());
        assertThrows(NoSuchElementException.class, s::pop);
    }
}
